Fix panic in Glob on empty subject string.

shurcooL opened the pull request:
Fixes #1.

Add tests.

Add test (in form of example) and benchmark for internal `globSequence` utility.

renfredxh commented on glob.go:
```diff
@@ -57,3 +62,14 @@ func Glob(pattern, subj string) bool {
+// I.e., it returns true for "", "*", "**", "***", etc., and false for all other patterns.
+func globSequence(pattern string) bool {
+	for i := 0; i < len(pattern); i++ {
+		if pattern[i] != '*' {
```
Should this be `pattern[i] != GLOB`? See how `const GLOB` is used in other places in this file.

shurcooL replied:
I wanted to do that, but couldn't.

`GLOB` is a `string` type. I need a `byte` here.

shurcooL replied:
I could do `GLOB[0]`, or declare `const globByte = '*'` instead, but I thought simply doing '*' would be okay.

renfredxh replied:
Got it, I saw your note about the performance implications of this as well. LGTM

ryanuber commented:
Welp that is pretty terrible! Thanks for finding this and fixing it.

I think we can actually do an even more simple fix though. On line 43..45, let's just remove the condition. If you are on the last chunk, you either have to match a trailing glob, or the suffix of the pattern. We should also switch the `return true` at the very end to `return false` and note that we should never reach it.

If we use the above approach we are actually just removing code which should speed us up a little bit, although we are talking about nanoseconds here. Let's definitely bring in your behavior tests with it.

Thanks again @shurcooL! Great work.

shurcooL commented:
Thanks for review, I will try that approach and update this PR.

shurcooL commented:
> We should also switch the `return true` at the very end to `return false` and note that we should never reach it.

If you're confident that it's not possible to ever reach it, I'd argue it's a better practice to do:

```Go
panic("unreachable")
```

With the implicit understanding that it should be impossible to get there, and you ever do, it's a bug or regression in the library that should be fixed.

E.g., something like [this](https://github.com/shurcooL/htmlg/blob/3da295397738f395c2ca5fd5570f01a9fcea3be3/htmlg.go#L46-L49), [this](https://github.com/shurcooL/vfsgen/blob/6e5fb4d6aad9620b57c54f045c7d738a64ff7563/test/test_vfsdata_test.go#L105-L106) or [this](https://github.com/shurcooL/vfsgen/blob/6e5fb4d6aad9620b57c54f045c7d738a64ff7563/test/test_vfsdata_test.go#L122-L123).

The motivation is that if you're 100% sure the logic should never get there in any situation, and it happens anyway, it's better to panic than to silently return incorrect results. It's also more readable and clear to other people that certain code is not a part of normal execution flow (otherwise they might think "how would the code ever get here? am I missing something?").

That said, it might be possible and even better to simply rearrange the logic so that a `panic("unreachable")` is not needed. For example, arrange so `case end` simply breaks out of the for loop, and have a final return statement that is executed. I will leave this to you.

shurcooL commented:
@ryanuber, I've implemented the change you suggested in a4a902f3562996a1cecd98ed1913f375fbba150c, PTAL.

ryanuber commented:
@shurcooL great suggestion for the panic, I agree. This looks great. Let's pull it in.

shurcooL commented:
All right, I've squashed into 1 commit and just added [2 more](https://github.com/ryanuber/go-glob/blob/87ce0b27039818fd8036c1d26e4d84e90e9c3648/glob_test.go#L34-L35) test cases to `TestEmptySubject`.

ryanuber commented:
:ship:

ryanuber commented:
Awesome, thanks again @shurcooL! I appreciate it.
